### PR TITLE
Close channel on handshake error with old version

### LINKED ISSUE #54337: Failed transport handshake between 6.8 and 7.6 nodes may throw a fatal AssertionError
Nodes send their transport handshakes with the version set to `Version.CURRENT.minimumCompatibilityVersion()`. For 6.8 nodes this is 5.6.0. When a 7.6 node receives a transport message, including a handshake, it responds using the same transport version. This means that a 7.6 node responding to a transport handshake claims to be sending a 5.6.0-compatible message. However, 7.6 nodes do not really know how to speak the 5.6.0 transport protocol. This mostly works since the format of a handshake response did not change between 5.6 and 6.0. However, the format of _error_ responses has changed, so an error response is not correctly deserialised. Worse, the incorrect deserialisation trips an assertion error that causes the 6.8 node to shut down:

```
[WARN ][org.elasticsearch.transport.TcpTransport] [REDACTED] exception caught on transport layer [Netty4TcpChannel{localAddress=/REDACTED, remoteAddress=/REDACTED}], clos
ing connection
java.lang.Exception: java.lang.AssertionError: illegal transport ID from node of version: 5.6.0 got: 1196 expected: 1
        at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.exceptionCaught(Netty4MessageChannelHandler.java:66) [transport-netty4-client-6.8.6.jar:6.8.6]
        at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:856) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:364) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1436) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1203) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1247) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
Caused by: java.lang.AssertionError: illegal transport ID from node of version: 5.6.0 got: 1196 expected: 1
        at org.elasticsearch.common.transport.TransportAddress.<init>(TransportAddress.java:86) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.transport.TransportAddress.<init>(TransportAddress.java:75) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamInput.readOptionalWriteable(StreamInput.java:777) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.ActionTransportException.<init>(ActionTransportException.java:41) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.RemoteTransportException.<init>(RemoteTransportException.java:45) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.readException(ElasticsearchException.java:306) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamInput.readException(StreamInput.java:799) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.handlerResponseError(TcpTransport.java:997) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:950) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:763) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:53) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        ... 31 more
[ERROR][org.elasticsearch.bootstrap.ElasticsearchUncaughtExceptionHandler] [REDACTED] fatal error in thread [...], exiting
java.lang.AssertionError: illegal transport ID from node of version: 5.6.0 got: 1196 expected: 1
        at org.elasticsearch.common.transport.TransportAddress.<init>(TransportAddress.java:86) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.transport.TransportAddress.<init>(TransportAddress.java:75) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamInput.readOptionalWriteable(StreamInput.java:777) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.ActionTransportException.<init>(ActionTransportException.java:41) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.RemoteTransportException.<init>(RemoteTransportException.java:45) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.readException(ElasticsearchException.java:306) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamInput.readException(StreamInput.java:799) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.handlerResponseError(TcpTransport.java:997) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:950) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:763) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:53) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[?:?]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) ~[?:?]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[?:?]
        at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[?:?]
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1436) ~[?:?]
        at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1203) ~[?:?]
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1247) ~[?:?]
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502) ~[?:?]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) ~[?:?]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[?:?]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[?:?]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) ~[?:?]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) ~[?:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) ~[?:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
```

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today a transport response uses the same wire format version as the
corresponding request. This mostly works since we mostly know we are
communicating with a node with a compatible version. TCP handshakes don't have
this guarantee since they use `Version.CURRENT.minimumCompatibilityVersion()`
to let us handshake with older nodes. This results in the strange situation of
a node of major version `N` responding to a node of major version `N-1` using a
wire format of version `N-2`.

We put extra effort into the longer BWC requirements for successful responses,
but we do not offer the same guarantees for error responses since they may be
rather complicated to serialize. This can result in the request sender misinterpreting
the response which may have unpredictable consequences.

Rather than strengthening the guarantees in this area, this commit simply logs
the exception and closes the connection on a handshake error with a node that
uses an incompatible wire format.

Closes #54337

### CODE REVIEW

**Comment by DaveCTurner** [comment]
The change to production code here is small and seems reasonable, but I really struggled to come up with sensible tests. Perhaps there's some appropriate seam that I'm missing? Anyway, this is what I ended up with, suggestions for alternative testing approaches are very welcome.

File: server/src/test/java/org/elasticsearch/transport/TransportServiceHandshakeTests.java [hunk 1]
```diff
@@ -190,4 +307,59 @@ public void testNodeConnectWithDifferentNodeId() {
...
+
+    private static InboundMessage unreadableHandshake(InboundMessage message) {
+        return new InboundMessage(message.getHeader(), false) {
+            @Override
+            public int getContentLength() {
+                return message.getContentLength();
+            }
+
+            @Override
+            public StreamInput openOrGetStreamInput() {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/transport/TransportServiceHandshakeTests.java:322]: I think we can do this way shorter as:

```java
            @Override
            public StreamInput openOrGetStreamInput() {
                final StreamInput streamInput = new InputStreamStreamInput(new InputStream() {
                    @Override
                    public int read() {
                        throw new ElasticsearchException("unreadable handshake");
                    }
                });
                streamInput.setVersion(message.getHeader().getVersion());
                return streamInput;
            }
```

**Review by original-brownbear — COMMENTED** [review]
Maybe we should just build this as a unit test for `InboundHandler`? The whole approach of testing this using the full transport services seems quite convoluted and in the end we're not mocking out the internals starting from where the message passes to `InboundHandler` anyway so it's not like we add coverage by doing that? (just a suggestion though, I realise it's a bit of extra effort to do this just for the sake of a having a nicer test)
Otherwise looks good I think from a quick read, just a suggestion on shortening things a little.

**Comment by DaveCTurner** [comment]
Ugh you're quite right this is now really just focussed on `InboundHandler` so let's just test that.

File: server/src/test/java/org/elasticsearch/transport/TransportServiceHandshakeTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/transport/TransportServiceHandshakeTests.java:322]: Thanks, I was pretty sure there was a way to do this by overriding a single method somewhere.

**Review by original-brownbear — APPROVED** [review]
LGTM (assuming CI is happy :)) Thanks for the extra iteration on the tests!

**Comment by DaveCTurner** [comment]
CI failure is unrelated, I opened https://github.com/elastic/elasticsearch/issues/57010.

@elasticmachine please run elasticsearch-ci/1

**Pull request merged**